src/bvals/cc/bvals_cc_in_one.cpp

```diff
@@ -0,0 +1,647 @@
+  ei = (nb.ni.ox1 < 0) ? (cellbounds.is(interior) + cn) : cellbounds.ie(interior);
+  sj = (nb.ni.ox2 > 0) ? (cellbounds.je(interior) - cn) : cellbounds.js(interior);
+  ej = (nb.ni.ox2 < 0) ? (cellbounds.js(interior) + cn) : cellbounds.je(interior);
+  sk = (nb.ni.ox3 > 0) ? (cellbounds.ke(interior) - cn) : cellbounds.ks(interior);
+  ek = (nb.ni.ox3 < 0) ? (cellbounds.ks(interior) + cn) : cellbounds.ke(interior);
+
+  // send the data first and later prolongate on the target block
+  // need to add edges for faces, add corners for edges
+  if (nb.ni.ox1 == 0) {
+    if (nb.ni.fi1 == 1)
+      si += pmb->block_size.nx1 / 2 - pmb->cnghost;
+    else
+      ei -= pmb->block_size.nx1 / 2 - pmb->cnghost;
+  }
+  if (nb.ni.ox2 == 0 && pmb->block_size.nx2 > 1) {
+    if (nb.ni.ox1 != 0) {
+      if (nb.ni.fi1 == 1)
+        sj += pmb->block_size.nx2 / 2 - pmb->cnghost;
+      else
+        ej -= pmb->block_size.nx2 / 2 - pmb->cnghost;
+    } else {
+      if (nb.ni.fi2 == 1)
+        sj += pmb->block_size.nx2 / 2 - pmb->cnghost;
+      else
+        ej -= pmb->block_size.nx2 / 2 - pmb->cnghost;
+    }
+  }
+  if (nb.ni.ox3 == 0 && pmb->block_size.nx3 > 1) {
+    if (nb.ni.ox1 != 0 && nb.ni.ox2 != 0) {
+      if (nb.ni.fi1 == 1)
+        sk += pmb->block_size.nx3 / 2 - pmb->cnghost;
+      else
+        ek -= pmb->block_size.nx3 / 2 - pmb->cnghost;
+    } else {
+      if (nb.ni.fi2 == 1)
+        sk += pmb->block_size.nx3 / 2 - pmb->cnghost;
+      else
+        ek -= pmb->block_size.nx3 / 2 - pmb->cnghost;
+    }
+  }
+}
+
+//----------------------------------------------------------------------------------------
+//! \fn size_t ResetAndRestrictSendBuffers(MeshData<Real> *md, bool cache_is_valid)
+//  \brief Resets boundary variable pointer (tbd if still required) and restricts
+//         cell centered variables if a cached version of boundary_info is used.
+//  \return The total number of buffers used in boundary_info
+
+size_t ResetAndRestrictSendBuffers(MeshData<Real> *md, bool cache_is_valid) {
+  Kokkos::Profiling::pushRegion("Reset boundaries");
+
+  size_t buffers_used = 0;
+  // reset buffer and count required buffers
+  for (int block = 0; block < md->NumBlocks(); block++) {
+    auto &rc = md->GetBlockData(block);
+    auto pmb = rc->GetBlockPointer();
+
+    int mylevel = pmb->loc.level;
+    for (auto &v : rc->GetCellVariableVector()) {
+      if (v->IsSet(parthenon::Metadata::FillGhost)) {
+        v->resetBoundary();
+        for (int n = 0; n < pmb->pbval->nneighbor; n++) {
+          parthenon::NeighborBlock &nb = pmb->pbval->neighbor[n];
+          auto *bd_var_ = v->vbvar->GetPBdVar();
+          if (bd_var_->sflag[nb.bufid] == parthenon::BoundaryStatus::completed) continue;
+          buffers_used += 1;
+
+          // Need to restrict here only if cached boundary_info is reused
+          // Otherwise restriction happens when the new boundary_info is created
```

Comment:
Unfortunately not with the current caching. More refactoring would be required to make that work.
The key design problem is that the caching only caches the buffer filling itself, which allows to keep the buffer filling kernel itself very simple.
One alternative approach (I already though about this) is to add more information to the cache, i.e, the `boundary_info`, so that the restriction could happen within the buffer filing kernel inline.
Another alternative is to completely separate the restriction into an own task and call that beforehand. This, in turn, would mean to introduce another cache where we store the indices and views to be restricted.

So bottom line: There's room for improvement but it would require some (potentially more extensive) refactoring.

Reply:
Understood. Good enough for now then. I think doing restriction as a separate task (in the long term) may be of value... if it's possible to disentangle. It may not be.